ext/afform/core/Civi/Api4/Subscriber/AutocompleteSubscriber.php

```diff
@@ -0,0 +1,77 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+namespace Civi\Api4\Subscriber;
+
+use Civi\Afform\FormDataModel;
+use Civi\API\Events;
+use Civi\Api4\Afform;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Preprocess api autocomplete requests
+ */
+class AutocompleteSubscriber implements EventSubscriberInterface {
+
+  /**
+   * @return array
+   */
+  public static function getSubscribedEvents() {
+    return [
+      'civi.api.prepare' => ['onApiPrepare', Events::W_MIDDLE],
+    ];
+  }
+
+  /**
+   * @param \Civi\API\Event\PrepareEvent $event
+   *   API preparation event.
+   */
+  public function onApiPrepare(\Civi\API\Event\PrepareEvent $event) {
+    $apiRequest = $event->getApiRequest();
+    if (is_object($apiRequest) && is_a($apiRequest, 'Civi\Api4\Generic\AutocompleteAction')) {
+      $formName = $apiRequest->getFormName();
+      if (!str_starts_with((string) $formName, 'afform:') || !strpos((string) $apiRequest->getFieldName(), ':')) {
+        return;
+      }
+      [$entityName, $fieldName] = explode(':', $apiRequest->getFieldName());
+      // Load afform only if user has permission
+      $afform = Afform::get()
+        ->addWhere('name', '=', str_replace('afform:', '', $formName))
+        ->addSelect('layout')
+        ->execute()->first();
+      if (!$afform) {
+        return;
+      }
+      $formDataModel = new FormDataModel($afform['layout']);
+      $entity = $formDataModel->getEntity($entityName);
+
+      // Look up the "type" fields (e.g. contact_type, activity_type_id, case_type_id, etc)
+      $typeFields = [];
+      if ($entity['type'] === 'Contact') {
+        $typeFields = ['contact_type', 'contact_sub_type'];
+      }
+      else {
+        $extends = array_column(\CRM_Core_BAO_CustomGroup::getCustomGroupExtendsOptions(), 'grouping', 'id');
+        $typeFields = (array) ($extends[$entity['type']] ?? NULL);
+      }
+      // If entity has a type set in the values, auto-apply that to filters
+      foreach ($typeFields as $typeField) {
+        if (!empty($entity['data'][$typeField])) {
+          $apiRequest->addFilter($typeField, $entity['data'][$typeField]);
+        }
+      }
+
+      $apiRequest->setCheckPermissions($entity['security'] !== 'FBAC');
+      $apiRequest->setSavedSearch($entity['fields'][$fieldName]['defn']['saved_search'] ?? NULL);
```

Comment:
(Interpreting/verbalizing) I've been reading with an eye to checking the security, and  believe the main concern is that the Afform visitor is only allowed to run the SavedSearch specified by the form.

The gist of `AutocompleteSubscriber` seems to be like: "*If the current user has permission to read the form (`Afform::get()` for `Autocompleteaction::$formName`), then inject `AutocompleteAction::$savedSearch` and `AutocompleteAction::$checkPermission`.*"

Which depends on a few things:

* The `Afform::get()` permission-check (L46-49) is a valid proxy for the `Afform::prefill()`/`Afform::submit()` permission-check. (*Anyone who can read the form is allowed to run the form.*)
    * *I'm not sure this is an essential characteristic of those APIs, but I'm guessing it's true.*
* This is the only way `AutocompleteAction::$savedSearch` is set.
    * *Here's where what I'm more curious about. If you call`AutocompleteAction`, you're allowed to pass in properties like `$savedSearch`. So if you pass in a dummy value (`$formName='not-afform:muaha'`), it'll short-circuit (L42 or L51) and leave whatever data was passed in for `$savedSearch`. That data may be `array` (*search-definition/preview-mode*). Doesn't that mean the caller can completely design their own search?

Reply:
@totten 
a) congratulations on getting your head 'round this so quickly. This PR is a beast.
b) You're assessment is correct. Anything coming from the client should not be trusted, so filters, formname & fieldname all must be validated. The `Afform::get()` is intended to check user permission to submit the form, and it's the same check used elsewhere by Afform to validate the same permission.
c) The `savedSearch` param is only accepted from the client as a string. API hooks would be allowed to inject an array instead of a string, but the client is not, due to the `@var string` annotation on that param in the `AutocompleteAction` class. I'll add a test to demonstrate that, because it's a good test to have.